pkg/typesystem/typesystem.go

```diff
@@ -742,50 +743,61 @@ func (t *TypeSystem) IsRelationWithRecursiveTTUAndAlgebraicOperations(objectType
 
 	w, ok := objRelNode.GetWeight(userType)
 	if !ok || w != graph.Infinite {
-		// 1. node[objectType#relation].weights[userType] = infinite
 		return false
 	}
 
 	edges, ok := t.authzWeightedGraph.GetEdgesFromNode(objRelNode)
-	if !ok || len(edges) != 1 {
-		// 2. node[objectType#relation] has only 1 edge
-		return false
-	}
-
-	edge := edges[0]
-	if edge.GetTo().GetLabel() != graph.UnionOperator {
-		// 2. node[objectType#relation] has 1 edge to an OR node
-		return false
-	}
-
-	unionNode := edge.GetTo()
-
-	edgesFromUnionNode, ok := t.authzWeightedGraph.GetEdgesFromNode(unionNode)
 	if !ok {
 		return false
 	}
 
-	tuplesetRelationKey := tuple.ToObjectRelationString(objectType, ttu.GetTupleset().GetRelation())
-	computedRelation := ttu.GetComputedUserset().GetRelation()
+	tuplesetRelation := tuple.ToObjectRelationString(objectType, ttu.GetTupleset().GetRelation())
+	computedRelationSuffix := "#" + ttu.GetComputedUserset().GetRelation()
 	recursiveTTUFound := false
 
-	for _, edge := range edgesFromUnionNode {
-		// find and validate the TTUEdge which is infinite (the one being processed at the current time)
-		if edge.GetEdgeType() == graph.TTUEdge &&
-			edge.GetConditionedOn() == tuplesetRelationKey && strings.HasSuffix(edge.GetTo().GetUniqueLabel(), "#"+computedRelation) {
+	for len(edges) != 0 {
+		innerEdges := make([]*graph.WeightedAuthorizationModelEdge, 0)
+
+		for _, edge := range edges {
 			w, ok := edge.GetWeight(userType)
-			if ok && w == graph.Infinite && edge.GetTo() == objRelNode {
-				recursiveTTUFound = true
+			if !ok {
+				// if the edge does not have a weight for the terminal type, we can skip it
 				continue
 			}
+			// edge is a set operator thus we have to inspect each node of the operator
+			if edge.GetEdgeType() == graph.RewriteEdge {
```

Comment:
want to double check whether we want to also validate whether we need to exclude computed userset (as they are also rewrite edge).  Or as they are added to the inner edge, they will be handled there

Reply:
Probably a good case to add to the typesystem_test.go

Reply:
Computed has it's own. See here: https://github.com/openfga/language/blob/main/pkg/go/graph/graph_edge.go#L14

Thus the only handler we need for `Computed` is here:
https://github.com/openfga/openfga/pull/2281/files#diff-cc97036cfe9a7c99fee528ae591f760c617ec84cc013c84dd01170657fc43ff4R793

Reply:
Understood.  Just want to confirm we do unit test this either here or at the matrix level so we can confirm typesystem handles this correctly.

Reply:
https://github.com/openfga/openfga/pull/2281/files#diff-5a15da103952b0c7e42a86c207025fe7a959bcc961cb79b2463aa85118cce83eR5797
covers it.